lib/rubycritic/source_control_systems/git/churn.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+module RubyCritic
+  module SourceControlSystem
+    class Git < Base
+      Stats = Struct.new(:count, :date)
+
+      class Renames
+        def initialize
+          @data = {}
+        end
+
+        def renamed(from, to)
+          current = current(to)
+          @data[from] = current
+        end
+
+        def current(name)
+          @data.fetch(name, name)
+        end
+      end
+
+      class Churn
+        def initialize
+          @renames = Renames.new
+          @date = nil
+          @stats = {}
+
+          call
+        end
+
+        def revisions_count(path)
+          stats(path).count
+        end
+
+        def date_of_last_commit(path)
+          stats(path).date
+        end
+
+        private
+
+        def call
+          Git
+            .git("log --all --follow --format='format:date:\t%ad' --name-status .")
+            .split("\n")
+            .reject { |line| line =~ /^$/ }
+            .each { |line| process_line(line) }
+        end
+
+        def process_line(line)
+          operation, rest = line.split("\t")
+
+          case operation
+          when /^date:$"/
+            process_date(rest)
+          when /^R/
+            process_rename(rest)
+          else
+            process_file(rest)
+          end
+        end
+
+        def process_date(date)
+          @date = DateTime.parse(date)
```

Comment:
Since the goal is to optimise the speed, maybe the date can be stored as a string, and only parsed+memoized when `date_of_last_commit` is called? Or as a last step of `call`.

The methods that rely on `Stats` might need to be updated as well. I'll leave it to your to see if it's worth it or not :)

Reply:
After reviewing the original code, it looks like it doesn't parse the date at all, and leaves it as a string, so that would make it match the original. I'll remove the parsing.